src/runtime.js

```diff
@@ -223,16 +221,24 @@ function variable_compute(variable) {
 
   const value0 = variable._value;
   const version = ++variable._version;
+  const inputs = variable._inputs;
+  const definition = variable._definition;
 
   // Lazily-constructed invalidation variable; only constructed if referenced as an input.
   let invalidation = null;
 
-  // If the variable doesn’t have any inputs, we can optimize slightly.
-  const promise = variable._promise = (variable._inputs.length
-      ? Promise.all(variable._inputs.map(variable_value)).then(define)
-      : new Promise(resolve => resolve(variable._definition.call(value0))))
+  // Wait for the previous definition to compute before recomputing.
+  const promise = variable._promise = variable._promise
+    .then(init, init)
```

Comment:
This change means that when a variable is redefined, it will wait for the old version of the variable to settle (fulfill or reject) before running. However, there are some important nuances here…

First, variables reject before computing if they are stale (`if (variable._version !== version)`). This means that if a variable is redefined multiple times, we won’t waste time computing the already-invalidated definitions. Second, new variables are created when code is edited rather than redefining existing variables. See Framework’s [`define`](https://github.com/observablehq/framework/blob/63c63e8764f01dc59adbf942cbdc0ca15964e7fe/src/client/main.js#L36), and note its call to `main.variable` to define a new variable whenever code is edited. This means that new code will _not_ wait for old code to settle before running — but only if the code is changed directly. Code that is downstream of changed code _will_ still wait for the old code to settle, which might be confusing.

We could fix the last point in Framework by redefining any downstream code as new variables if it proves to be confusing. As a workaround, you can reload the page if you don’t wait for old code to run.

Reply:
A slight delay is OK, but it is definitely confusing when the unchanged code is not settling at all.

A typical example is when I make a mistake writing a cell as a Promise that never resolves; fixing the mistake in the code block does not fix the state of the page.

Reply:
So what do you suggest we do?

Reply:
After playing a bit more with it, only the programmer error case mentioned above can be seen as a regression, ending up in a situation where you have to reload the page to resync the tree of variables with your code — whereas in main fixing the code would fix it.

As for suggestion, I wish I had an idea :) “Redefining any downstream code as new variables”, as you mentioned, maybe with an additional timeout(?).

But the additional complexity might not be worth it. And this is such a big win that if we have to choose I prefer this version.

Reply:
We could make it so that calls to variable.define cause the variable and any downstream variables not to wait, such that we’d only end up waiting in the case of generators yielding a new value. I can look to see how hard that will be.